pkg/controller/pagerdutyintegration/clusterdeployment_created.go

```diff
@@ -23,121 +23,161 @@ import (
 
 	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1"
 	"github.com/openshift/pagerduty-operator/config"
+	pagerdutyv1alpha1 "github.com/openshift/pagerduty-operator/pkg/apis/pagerduty/v1alpha1"
 	"github.com/openshift/pagerduty-operator/pkg/kube"
 	"github.com/openshift/pagerduty-operator/pkg/localmetrics"
 	pd "github.com/openshift/pagerduty-operator/pkg/pagerduty"
 	"github.com/openshift/pagerduty-operator/pkg/utils"
 
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
-	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
-func (r *ReconcileClusterDeployment) handleCreate(request reconcile.Request, instance *hivev1.ClusterDeployment) (reconcile.Result, error) {
-	if !instance.Spec.Installed {
+func (r *ReconcilePagerDutyIntegration) handleCreate(pdi *pagerdutyv1alpha1.PagerDutyIntegration, cd *hivev1.ClusterDeployment) error {
+	var (
+		// secretName is the name of the Secret deployed to the target
+		// cluster, and also the name of the SyncSet that causes it to
+		// be deployed.
+		secretName string = config.Name(pdi.Spec.ServicePrefix, cd.Name, config.SecretSuffix)
+
+		// configMapName is the name of the ConfigMap containing the
+		// SERVICE_ID and INTEGRATION_ID
+		configMapName string = config.Name(pdi.Spec.ServicePrefix, cd.Name, config.ConfigMapSuffix)
+
+		// There can be more than one PagerDutyIntegration that causes
+		// creation of resources for a ClusterDeployment, and each one
+		// will need a finalizer here. We add a suffix of the CR
+		// name to distinguish them.
+		finalizer string = "pd.managed.openshift.io/" + pdi.Name
+	)
+
+	if !cd.Spec.Installed {
 		// Cluster isn't installed yet, return
-		return reconcile.Result{}, nil
+		return nil
 	}
 
-	if utils.HasFinalizer(instance, config.OperatorFinalizer) == false {
-		utils.AddFinalizer(instance, config.OperatorFinalizer)
-		err := r.client.Update(context.TODO(), instance)
+	if utils.HasFinalizer(cd, finalizer) == false {
+		utils.AddFinalizer(cd, finalizer)
+		err := r.client.Update(context.TODO(), cd)
```

Comment:
When you say another thread, do you mean another thread in this process (concurrent reconciles or a different controller or a different instance of this controller), or in another operator such as hive? My understanding is that there will only be one reconcile at a time in this controller (we don't override the `MaxConcurrentReconciles` option when the controller is created, and the default is 1). Also, even if there are multiple instances of this operator deployed, only one will be active, after [acquiring the lock](https://github.com/openshift/pagerduty-operator/blob/master/cmd/manager/main.go#L100).

The reason I'm bringing this up is that I'm hesitant to change this, as it would require changing other things -- this function now just returns an error value, and relies on the caller to handle that (either continue or return an empty reconcile result & the error, etc). While I _can_ change this here, I'm hesitant to change too much logic that existed before the changes in this pull request, especially when I'm not from the owning team). WDYT?

Reply:
Agree, any changes to the CR you're watching must result in existing the reconcile.  Another event is triggered for that CR and will be processed by all watching controllers.  You will have race conditions if you continue.

Reply:
I still don't see where a race condition could occur here (did the possibility of it exist before this change?), but I'll update now in any case.

Reply:
Fixed in latest commit :)